File: src/test/java/com/team10/backend/domain/image/service/ImageUploadServiceTest.java
package com.team10.backend.domain.image.service;

import com.team10.backend.domain.image.dto.ImageUploadResponse;
import com.team10.backend.domain.image.dto.PresignedUrlRequest;
import com.team10.backend.domain.image.dto.PresignedUrlResponse;
import com.team10.backend.global.exception.BusinessException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.test.util.ReflectionTestUtils;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.S3Presigner;
import software.amazon.awssdk.services.s3.presigner.model.PresignedPutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest;

import java.net.URI;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class ImageUploadServiceTest {

    private S3Client s3Client;
    private S3Presigner s3Presigner;
    private ImageUploadService imageUploadService;

    @BeforeEach
    void setUp() {
        s3Client = mock(S3Client.class);
        s3Presigner = mock(S3Presigner.class);
        imageUploadService = new ImageUploadService(s3Client, s3Presigner);
        ReflectionTestUtils.setField(imageUploadService, "bucket", "team10-images-dev-test");
        ReflectionTestUtils.setField(imageUploadService, "region", "ap-northeast-2");
    }

    @Test
    void uploadImageReturnsS3Url() {
        MockMultipartFile file = new MockMultipartFile(
                "file",
                "cake.jpg",
                "image/jpeg",
                "test-image".getBytes()
        );

        ImageUploadResponse response = imageUploadService.upload(file, "products");

        assertThat(response.imageUrl)
                .startsWith("https://team10-images-dev-test.s3.ap-northeast-2.amazonaws.com/products/")
                .endsWith(".jpg");
        verify(s3Client).putObject(any(PutObjectRequest.class), any(software.amazon.awssdk.core.sync.RequestBody.class));
    }

    @Test
    void createPresignedUrlReturnsUploadUrlAndImageUrl() throws Exception {
        PresignedPutObjectRequest presignedRequest = mock(PresignedPutObjectRequest.class);
        when(presignedRequest.url()).thenReturn(URI.create("https://presigned-upload.test/upload").toURL());
        when(s3Presigner.presignPutObject(any(PutObjectPresignRequest.class))).thenReturn(presignedRequest);

        PresignedUrlResponse response = imageUploadService.createPresignedUrl(
                new PresignedUrlRequest("cake.jpg", "image/jpeg", "products")
        );

        assertThat(response.uploadUrl).isEqualTo("https://presigned-upload.test/upload");
        assertThat(response.imageUrl)
                .startsWith("https://team10-images-dev-test.s3.ap-northeast-2.amazonaws.com/products/")
                .endsWith(".jpg");
        verify(s3Presigner).presignPutObject(any(PutObjectPresignRequest.class));
    }

    @Test
    void createPresignedUrlRejectsNonImageContentType() {
        PresignedUrlRequest request = new PresignedUrlRequest("memo.txt", "text/plain", "products");

        assertThatThrownBy(() -> imageUploadService.createPresignedUrl(request))
                .isInstanceOf(BusinessException.class);
    }

    @Test
    void uploadRejectsNonImageFile() {
        MockMultipartFile file = new MockMultipartFile(
                "file",
                "memo.txt",
                "text/plain",
                "not-image".getBytes()
        );

        assertThatThrownBy(() -> imageUploadService.upload(file, "products"))
                .isInstanceOf(BusinessException.class);
    }

    @Test
    void deleteImageDeletesS3Object() {
        String imageUrl = "https://team10-images-dev-test.s3.ap-northeast-2.amazonaws.com/feeds/test-image.jpg";

        imageUploadService.delete(imageUrl);

        verify(s3Client).deleteObject(any(DeleteObjectRequest.class));
    }

    @Test
    void deleteRejectsExternalImageUrl() {
        String imageUrl = "https://other-bucket.s3.ap-northeast-2.amazonaws.com/feeds/test-image.jpg";

        assertThatThrownBy(() -> imageUploadService.delete(imageUrl))
                .isInstanceOf(BusinessException.class);
        verify(s3Client, never()).deleteObject(any(DeleteObjectRequest.class));
    }

    @Test
    void deleteIfManagedSkipsExternalImageUrl() {
        String imageUrl = "https://other-bucket.s3.ap-northeast-2.amazonaws.com/feeds/test-image.jpg";

        imageUploadService.deleteIfManaged(imageUrl);

        verify(s3Client, never()).deleteObject(any(DeleteObjectRequest.class));
    }
}

File: src/test/kotlin/com/team10/backend/domain/image/service/ImageUploadServiceTest.kt
package com.team10.backend.domain.image.service

import com.team10.backend.domain.image.dto.PresignedUrlRequest
import com.team10.backend.global.exception.BusinessException
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.springframework.test.util.ReflectionTestUtils
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest
import software.amazon.awssdk.services.s3.presigner.S3Presigner
import software.amazon.awssdk.services.s3.presigner.model.PresignedPutObjectRequest
import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest
import java.net.URI

internal class ImageUploadServiceTest {
    private lateinit var s3Client: S3Client
    private lateinit var s3Presigner: S3Presigner
    private lateinit var imageUploadService: ImageUploadService

    @BeforeEach
    fun setUp() {
        s3Client = Mockito.mock(S3Client::class.java)
        s3Presigner = Mockito.mock(S3Presigner::class.java)
        imageUploadService = ImageUploadService(s3Client, s3Presigner)
        ReflectionTestUtils.setField(imageUploadService, "bucket", "team10-images-dev-test")
        ReflectionTestUtils.setField(imageUploadService, "region", "ap-northeast-2")
    }

    @Test
    @Throws(Exception::class)
    fun createPresignedUrlReturnsUploadUrlAndImageUrl() {
        val presignedRequest = Mockito.mock(PresignedPutObjectRequest::class.java)
        Mockito.`when`(presignedRequest.url())
            .thenReturn(URI.create("https://presigned-upload.test/upload").toURL())
        Mockito.`when`(
            s3Presigner.presignPutObject(
                ArgumentMatchers.any(
                    PutObjectPresignRequest::class.java
                )
            )
        ).thenReturn(presignedRequest)

        val response = imageUploadService.createPresignedUrl(
            PresignedUrlRequest("cake.jpg", "image/jpeg", "products")
        )

        Assertions.assertThat(response.uploadUrl).isEqualTo("https://presigned-upload.test/upload")
        Assertions.assertThat(response.imageUrl)
            .startsWith("https://team10-images-dev-test.s3.ap-northeast-2.amazonaws.com/products/")
            .endsWith(".jpg")
        Mockito.verify(s3Presigner)
            .presignPutObject(ArgumentMatchers.any(PutObjectPresignRequest::class.java))
    }

    @Test
    fun createPresignedUrlRejectsNonImageContentType() {
        val request = PresignedUrlRequest("memo.txt", "text/plain", "products")

        Assertions.assertThatThrownBy { imageUploadService.createPresignedUrl(request) }
            .isInstanceOf(BusinessException::class.java)
    }

    @Test
    fun deleteImageDeletesS3Object() {
        val imageUrl = "https://team10-images-dev-test.s3.ap-northeast-2.amazonaws.com/feeds/test-image.jpg"

        imageUploadService.delete(imageUrl)

        Mockito.verify(s3Client)
            .deleteObject(ArgumentMatchers.any(DeleteObjectRequest::class.java))
    }

    @Test
    fun deleteRejectsExternalImageUrl() {
        val imageUrl = "https://other-bucket.s3.ap-northeast-2.amazonaws.com/feeds/test-image.jpg"

        Assertions.assertThatThrownBy { imageUploadService.delete(imageUrl) }
            .isInstanceOf(BusinessException::class.java)
        Mockito.verify(s3Client, Mockito.never()).deleteObject(
            ArgumentMatchers.any(
                DeleteObjectRequest::class.java
            )
        )
    }

    @Test
    fun deleteIfManagedSkipsExternalImageUrl() {
        val imageUrl = "https://other-bucket.s3.ap-northeast-2.amazonaws.com/feeds/test-image.jpg"

        imageUploadService.deleteIfManaged(imageUrl)

        Mockito.verify(s3Client, Mockito.never()).deleteObject(
            ArgumentMatchers.any(
                DeleteObjectRequest::class.java
            )
        )
    }
}
